In [10]:
from sklearn.metrics import pairwise_distances

from pipelines.shared.string_utils import split_camel_case
# Setup

%load_ext autoreload
%autoreload 2

import os

import torch
import numpy as np
from umap import UMAP
from torch.nn.functional import normalize
from dotenv import load_dotenv
from pandas import read_json
from sklearn.cluster import HDBSCAN
from transformers import AutoModel, AutoTokenizer

from pipelines.multi_service.mean_pooling import mean_pooling

load_dotenv()

print(f"HuggingFace cache directory is {os.environ.get('HF_HOME')}")


def should_be_skipped(class_name: str):
    return (
        class_name.endswith("Test")
        or class_name.endswith("Tests")
        or class_name.endswith("IT")
    )

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload
HuggingFace cache directory is /media/martin/BigData/datasets/cache/huggingface/


In [82]:
import pandas as pd

# MODEL = "google-bert/bert-base-uncased"
# MODEL = "google-bert/bert-base-cased"
GENERIC_WORDS = [
    "get",
    "set",
    "find by",
    "exists by",
    "exists",
    "update",
    "delete",
    "delete by",
    "create",
    "get all",
    "create",
    "insert",
    "insert into",
    "find all",
    "find all by",
]
MODEL = "microsoft/codebert-base"

codeBertTokenizer = AutoTokenizer.from_pretrained(MODEL)
codeBertModel = AutoModel.from_pretrained(MODEL)

# bertTokenizer = AutoTokenizer.from_pretrained("google-bert/bert-base-uncased")
# bertModel = AutoModel.from_pretrained("google-bert/bert-base-uncased")

# input_file = "/media/martin/BigData/datasets/pa165-git/01/multi-service.jsonl"
# output_file = "/media/martin/BigData/datasets/pa165-git/01/multi-service-result.jsonl"

input_file = (
    "/media/martin/BigData/datasets/github-projects/klaw/multi-service-labeled.jsonl"
)
output_file = (
    "/media/martin/BigData/datasets/github-projects/klaw/final-with-centroids.jsonl"
)

# input_file = "/media/martin/BigData/datasets/github-projects/cas/multi-service.json"
# output_file = "/media/martin/BigData/datasets/github-projects/cas/multi-service-result-codebert-combined.jsonl"

# input_file = "/media/martin/BigData/datasets/pa165-git/multi-service-filtered-labeled-evaluated.jsonl"
# output_file = "/media/martin/BigData/datasets/pa165-git/multi-service-with-distances.jsonl"

data_frame = read_json(input_file, lines=True)

# origin_property = "methodSignature"
# evaluation_property = "methodSignatures"


def extract_property(methods: list[dict[str, str]], property: str):
    return list(map(lambda m: m[property], methods))


def remove_words(text: str):
    for word in GENERIC_WORDS:
        if text.startswith(word):
            text = text.removeprefix(word).strip()

    return text


def evaluate(inputs: list[dict[str, str]]):
    mean_pooled_signature = get_embeddings(extract_property(inputs, "signature"))
    method_names = extract_property(inputs, "name")
    method_name_split = list(
        map(lambda name: remove_words(" ".join(split_camel_case(name))), method_names)
    )
    mean_names = get_embeddings(method_name_split)

    signature_normalized = normalize(mean_pooled_signature, p=2, dim=1)
    name_normalized = normalize(mean_names, p=2, dim=1)

    combined = torch.cat((signature_normalized, name_normalized), dim=-1)
    mean_pooled_numpy = combined.cpu().numpy()

    # pca = PCA(n_components=0.95, random_state=42)
    # pca_reduced = pca.fit_transform(mean_pooled_numpy)

    umap = UMAP(
        n_neighbors=5,
        min_dist=0.0,
        n_components=15,
        metric="cosine",
        random_state=42,
        init="random",
    )
    umap_reduced = umap.fit_transform(mean_pooled_numpy)
    # gap_df = calculate_gap_statistic(
    #     pca_reduced, n_refs=5, max_k=math.ceil(math.sqrt(len(inputs)))
    # )

    clustering_alg = HDBSCAN(
        min_samples=1,
        min_cluster_size=3,
        metric="cosine",
        allow_single_cluster=True,
        store_centers="medoid",
    )

    labels = clustering_alg.fit_predict(umap_reduced)

    unique_labels = set(labels)
    number_of_clusters = len(unique_labels) - (1 if -1 in unique_labels else 0)
    noise_count = list(labels).count(-1)
    noise_ratio = noise_count / len(labels)
    # return (get_optimal_k_programmatically(gap_df)["optimal_k"], pca_reduced)

    print(labels)
    print(clustering_alg.medoids_)
    medoids = clustering_alg.medoids_
    unique_clusters = unique_labels - {-1}
    avg_distance = 0.0
    normalized_avg_distance = 0.0
    if len(unique_clusters) > 1:
        # centroids = []
        # for label in unique_clusters:
        #     centroids.append(np.mean(umap_reduced[labels == label], axis=0))

        if len(unique_clusters) > 1:
            distances = pairwise_distances(medoids, metric="cosine")
            print(distances)
            tri_offset = np.triu_indices(distances.shape[0], k=1)
            unique_distances = distances[tri_offset]

            avg_distance = np.mean(unique_distances)

            # distances = pairwise_distances(centroids, metric="euclidean")
            # tri_offset = np.triu_indices(distances.shape[0], k=1)
            # unique_distances = distances[tri_offset]
            #
            # # 5. Calculate Metrics
            # avg_distance = np.mean(unique_distances)
            # max_distance = np.max(unique_distances)
            # normalized_avg_distance = avg_distance / max_distance

    if all(x == -1 for x in labels):
        number_of_clusters = -1

    return (
        number_of_clusters,
        noise_ratio,
        umap_reduced,
        labels,
        avg_distance,
        normalized_avg_distance,
    )


def get_embeddings(inputs, tokenizer=codeBertTokenizer, model=codeBertModel):
    tokenized_inputs = tokenizer(
        inputs, return_tensors="pt", padding=True, truncation=True, max_length=256
    )
    with torch.no_grad():
        output = model(**tokenized_inputs)
    attention_mask = tokenized_inputs["attention_mask"]
    mean_pooled = mean_pooling(output, attention_mask)
    return mean_pooled


data_frame["optimal_k"] = pd.Series()
data_frame["noise_ratio"] = pd.Series()
data_frame["cluster"] = pd.Series()
data_frame["avg_dist"] = pd.Series(dtype="float64")
results = []
for index, row in data_frame.iterrows():
    class_name = row["fullyQualifiedName"]
    methods = row["methods"]

    if len(methods) < 10:
        print(f"Skipped small class {class_name}, number of methods {len(methods)}")
        continue

    if should_be_skipped(class_name):
        print(f"Skipped Test class {class_name}, number of methods {len(methods)}")
        continue

    optimal_k, noise_ratio, pca_reduced, labels, avg_dist, normalized_avg_distance = (
        evaluate(methods)
    )

    # kmeans = KMeans(n_clusters=optimal_k, random_state=42)
    # cluster = kmeans.fit_predict(pca_reduced)
    data_frame.loc[index, "optimal_k"] = optimal_k
    data_frame.loc[index, "avg_dist"] = avg_dist * 100
    data_frame.loc[index, "noise_ratio"] = noise_ratio
    # data_frame.at[index, "cluster"] = cluster.tolist()
    data_frame.at[index, "cluster"] = labels

    results.append(f"{class_name}: {optimal_k}, {labels}, dist: {avg_dist:.4f}")

for result in results:
    print(result)

data_frame.to_json(output_file, lines=True, orient="records")

Skipped small class io.aiven.klaw.clusterapi.KafkaClusterApiApplication, number of methods 1
Skipped small class io.aiven.klaw.clusterapi.models.ClusterTopicRequest, number of methods 0
Skipped small class io.aiven.klaw.clusterapi.models.ClusterSchemaRequest, number of methods 0
Skipped small class io.aiven.klaw.clusterapi.models.ApiResponse, number of methods 2
Skipped small class io.aiven.klaw.clusterapi.models.ServiceAccountDetails, number of methods 0
Skipped small class io.aiven.klaw.clusterapi.models.SchemaCompatibilityCheckResponse, number of methods 0
Skipped small class io.aiven.klaw.clusterapi.models.TopicConfig, number of methods 0
Skipped small class io.aiven.klaw.clusterapi.models.ClusterSchemaCacheResetRequest, number of methods 0
Skipped small class io.aiven.klaw.clusterapi.models.LoadTopicsResponse, number of methods 0
Skipped small class io.aiven.klaw.clusterapi.config.JwtRequestFilter, number of methods 0
Skipped small class io.aiven.klaw.clusterapi.models.AivenAclRes

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 0 0 1 1 1 1 1 0 0 0 0]
[[ 4.09609032  2.3872335   2.83578348  3.48891807  1.91486609  7.3935833
   0.89180213  6.41921043  4.17595482  3.53075647  2.55974913  6.51638031
   4.22281218  7.32380342  1.61292815]
 [ 2.51680112 -0.29010129  4.75836039  4.98439217  5.11956978  7.56398153
   0.41773781  3.01133871  2.03985357  1.88203907  3.52684569  3.51438451
   4.64435387  8.19583988 -1.0852598 ]]
[[0.         0.11080344]
 [0.11080344 0.        ]]
Skipped small class io.aiven.klaw.clusterapi.utils.MetricsUtils, number of methods 1
Skipped small class io.aiven.klaw.clusterapi.utils.AdminClientProperties, number of methods 0


/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 0 1 1 1 0 0 0]
[[ 6.98860693 -3.07591772  6.69377184 13.20521736  4.78230429 14.06496143
   0.52673686  9.55979347 12.14513206  5.62540436 10.23738003  1.33115554
   5.229321    7.07524967  5.95437002]
 [ 7.00570059 -2.32967925  6.64405441 13.88626003  5.08502531 12.73556232
   0.42557588  8.61315346 11.58621883  5.58508301 10.07354736  1.92013144
   4.19719553  6.57941341  5.40854359]]
[[0.         0.00244781]
 [0.00244781 0.        ]]


/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 1 0 0 0 0 0 1 1]
[[ 0.76603538  1.10615647  6.12567139  5.61887407  4.54169178  1.30250931
   1.8362782   3.74048257  0.71444446  1.84510124  8.28319168 12.8821888
   5.43736076  0.73873913  6.4252038 ]
 [ 1.15311897  1.76392925  5.75723505  5.25851345  5.68886614  0.49070951
   2.17301154  3.72373176  1.5984019   2.81071591  8.40403366 11.84988213
   4.69454527  1.02238369  6.00668716]]
[[0.         0.00759187]
 [0.00759187 0.        ]]
Skipped small class io.aiven.klaw.clusterapi.controller.TopicContentsController, number of methods 1
Skipped small class io.aiven.klaw.clusterapi.services.JwtTokenUtilService, number of methods 3
Skipped small class io.aiven.klaw.clusterapi.controller.KafkaConnectController, number of methods 8
Skipped small class io.aiven.klaw.clusterapi.controller.MetricsApiController, number of methods 1
Skipped small class io.aiven.klaw.clusterapi.services.ApacheKafkaAclService, number of methods 3
Skipped small class io.aiven.klaw.clusterapi.services.Apache

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[ 0 10  6  6  6  6  2  2  2  1  2  6  8 14 14  0  7 14  3  3  3 14 13 -1
 -1  9  9  9  9 -1 14 13 14 14 14  6 15 -1 16  0  1  1  1  0  1 10 10 10
  7 13 15 -1 15 16 -1 11 13  7  4 11  4 11  1  1  1 12 12 12 16  4 11 16
 -1  8  8 16 -1  7  8 -1  7  3  3 12 14  5  1  5  5  3  2  8]
[[ 3.53525329  7.77355623  8.46909428  4.23975754  5.36336946 -0.47078526
   3.76378298  5.20163488  3.34544683  4.88004732  2.73604512  6.28252554
   0.05929476  1.65966141  4.08057165]
 [ 3.4363358   8.05198479  8.24752903  4.07635832  5.76957226 -0.26060987
   3.88884187  5.3614974   3.28180456  4.99512959  2.48056936  6.54213476
   0.33882251  1.80210519  4.43505192]
 [ 3.28331089  7.51414108  7.66008615  4.54106951  4.65125179  0.26017937
   4.53919601  5.59554338  3.83795762  5.35851526  3.08282757  6.45270348
   0.45624131  1.38169622  3.96927667]
 [ 3.235075    7.45228624  7.49705935  4.77393103  4.44185352  0.38792145
   4.7766242   5.85581827  4.0880065   5.6314559   3.39917207  6.46155405
   0.49804

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 0 0 0 0 0 0 0]
[[5.54049492 4.23387718 4.40681696 5.0666852  2.68079925 6.44647312
  5.22454071 1.69710183 3.4732213  2.16157532 1.90218663 9.86189556
  6.29399347 4.24175215 2.92052197]]
Skipped small class io.aiven.klaw.constants.TestConstants, number of methods 0
Skipped small class io.aiven.klaw.SecurityConfigNoSSOIT, number of methods 7
Skipped small class io.aiven.klaw.controller.UserTeamsControllerTest, number of methods 3
Skipped small class io.aiven.klaw.config.ManageDatabaseTest, number of methods 7
Skipped small class io.aiven.klaw.auth.KwAuthenticationSuccessHandlerTest, number of methods 7
Skipped small class io.aiven.klaw.controller.ServerConfigControllerTest, number of methods 2
Skipped small class io.aiven.klaw.controller.SchemaRegistrySyncControllerTest, number of methods 5
Skipped small class io.aiven.klaw.controller.SchemaRegistryControllerTest, number of methods 6
Skipped Test class io.aiven.klaw.service.ServerConfigServiceTest, number of methods 12
Skipped T

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 0 0 1 1 0 1 0]
[[ 9.56680202  2.06872129  2.53225684  5.7735815   5.85452652  6.74493504
   4.16865921  1.17011571  5.24748039  4.11902142  2.12300992  3.57254243
   5.67417622  5.80192423  3.18707681]
 [10.01821995  2.3415556   1.89567494  5.80222178  5.17021084  6.41586399
   4.90709162  0.88151795  4.82705069  3.75646234  1.96264362  3.5847609
   5.50559807  5.78294039  2.18085814]]
[[0.         0.00430326]
 [0.00430326 0.        ]]
Skipped small class io.aiven.klaw.model.CaptchaResponse, number of methods 0
Skipped small class io.aiven.klaw.dao.OperationalRequest, number of methods 0
Skipped small class io.aiven.klaw.dao.UserInfo, number of methods 0
Skipped small class io.aiven.klaw.model.TopicInfo, number of methods 0
Skipped small class io.aiven.klaw.dao.RegisterUserInfo, number of methods 0
Skipped small class io.aiven.klaw.model.NotificationModel, number of methods 0
Skipped small class io.aiven.klaw.model.TopicBaseConfig, number of methods 0
Skipped small class io.aive

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[1 1 2 2 2 1 1 1 1 1 0 1 0 0 0 0 0 0 1]
[[ 7.83825397  4.34352589  8.02275372  5.9254961   0.37520251  3.20396972
   6.85703611  7.44368124  7.93626118  3.64661002  9.52919865  3.01326442
   5.37076569  6.17055082  3.40947461]
 [ 2.24996185  2.56919146 10.30826664  5.92745876  1.97059298  2.19966221
   5.61918211  3.94719815  6.62513924  4.91091394  9.14472675  5.29354191
   3.20577145  7.98768854  4.19537115]
 [ 5.0375123   3.38088751  8.85919285  5.79301453  1.2547524   2.62891245
   6.21149302  5.66330671  7.5243063   4.06736183  9.62652779  4.44089413
   4.08961964  6.93111849  3.56036067]]
[[0.         0.07021015 0.01667263]
 [0.07021015 0.         0.01983183]
 [0.01667263 0.01983183 0.        ]]
Skipped small class io.aiven.klaw.dao.ServiceAccounts, number of methods 0
Skipped small class io.aiven.klaw.model.TopicConfigEntry, number of methods 0
Skipped small class io.aiven.klaw.model.requests.TopicCreateRequestModel, number of methods 0
Skipped small class io.aiven.klaw.dao.KwMe

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 0 0 0 0 0 0 0 0 0]
[[7.34001112 4.40750933 2.56524944 9.20660591 3.0243597  9.42576313
  6.9810853  3.70944381 7.6920886  4.49012136 6.10962725 2.43487263
  3.48507762 7.08714962 4.0470624 ]]
Skipped small class io.aiven.klaw.model.cluster.consumergroup.OffsetDetails, number of methods 0
Skipped small class io.aiven.klaw.model.response.RequestsOperationTypeCount, number of methods 0
Skipped small class io.aiven.klaw.controller.AnalyticsController, number of methods 5
Skipped small class io.aiven.klaw.model.response.SchemaDetailsPerEnv, number of methods 0
Skipped small class io.aiven.klaw.model.cluster.consumergroup.ResetConsumerGroupOffsetsRequest, number of methods 0
Skipped small class io.aiven.klaw.model.response.TopicDetailsPerEnv, number of methods 0
Skipped small class io.aiven.klaw.model.response.TopicRequestsResponseModel, number of methods 0
Skipped small class io.aiven.klaw.model.response.ClusterInfo, number of methods 0
Skipped small class io.aiven.klaw.model.respons

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[1 3 4 2 2 3 4 2 4 1 2 2 3 2 0 3 0 0 0 0 0 0 0 1]
[[ 4.72812605  6.38609028  3.66127419  1.25870633  3.04936242  4.90738344
   6.13174486  7.82585621  7.63830757  3.28711081  3.86006927  8.82899952
   3.6162858   4.44918346  1.19523335]
 [ 9.51873779  6.95954752  4.41202831 -2.24602222  8.60700226  9.16551208
   9.44582939  4.63365555 -0.6715945   3.12551904  3.14056063  7.95409727
   1.89543343  5.10396528  3.85721564]
 [10.23455524  7.8205905   5.57448101 -2.27026939  7.72226763  7.72088718
   9.51127815  4.32909441 -1.38015568  3.89721608  3.5276475   7.21302605
   2.58289242  5.22415924  3.44669509]
 [ 9.63264751  7.16420841  4.68344212 -2.05431652  8.27248192  8.79351044
   9.47677708  4.70891428 -1.06068933  3.38928223  3.49529719  7.61939049
   2.22182798  5.26253986  3.57806802]
 [ 9.94677734  7.76981735  5.18369675 -2.09684348  7.78386784  8.26844597
   9.57462978  4.68265533 -1.35626328  3.59445024  3.63733315  7.41880369
   2.32018447  5.12996054  3.43864465]]
[[0.         0

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[1 1 2 2 2 2 0 0 1 0 2]
[[ 4.77302074 -0.68359244  8.98859787  5.89465237  3.26313591  7.99542427
   9.27475166  1.96217108  1.02971315  2.88422322 -1.88813126  1.3291049
   2.88706613  6.6625042   6.7691884 ]
 [ 4.81107378 -0.72382796 10.47166538  4.98419809  2.52336335  8.30862999
   9.11687565  2.46839595  1.10336828  3.11354256 -1.35036409  1.34086204
   2.75274134  6.83148718  5.73863602]
 [ 5.40379143 -1.28299177  9.98093033  4.98169851  2.93226719  7.72937489
   9.89431763  2.60093498  0.90133756  2.88153338 -1.4066937   0.88736188
   3.25287938  7.06389046  6.30737925]]
[[0.         0.00639278 0.00486192]
 [0.00639278 0.         0.0032537 ]
 [0.00486192 0.0032537  0.        ]]
Skipped small class io.aiven.klaw.model.response.RequestEntityStatusCount, number of methods 0
Skipped small class io.aiven.klaw.dao.migration.MigrateData2x8x0, number of methods 1
Skipped small class io.aiven.klaw.model.charts.Options, number of methods 0
Skipped small class io.aiven.klaw.model.response.

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 0 0 0 1 1 0 0 1 1 0]
[[6.41283655 3.83484554 5.27284098 1.09556174 6.21603155 3.0416708
  1.82925522 8.33112717 1.23758674 7.47240448 5.92304707 3.61381412
  3.89594364 4.87066793 4.54327202]
 [6.11531448 5.07693672 4.81918287 2.39589858 6.18115854 2.66217995
  1.5320183  7.77344799 2.51093912 8.05304623 6.45329714 2.23193312
  4.2460866  5.19803381 4.16311884]]
[[0.         0.01136303]
 [0.01136303 0.        ]]
Skipped small class io.aiven.klaw.model.charts.TeamOverview, number of methods 0
Skipped small class io.aiven.klaw.service.BaseOverviewService, number of methods 0
Skipped small class io.aiven.klaw.model.charts.Title, number of methods 0
Skipped small class io.aiven.klaw.controller.CoralController, number of methods 1
Skipped small class io.aiven.klaw.model.charts.YAx, number of methods 0
Skipped small class io.aiven.klaw.config.ConfigUtils, number of methods 0
Skipped small class io.aiven.klaw.repository.KwEntitySequenceRepo, number of methods 3
Skipped small class io.a

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 3 3 3 3 3 0 1 2 1 2 2 1 1 1 0]
[[ 2.67460775  2.89235544  7.31051779 11.47279072  7.70037556 -1.69169378
   0.618146    2.80139852  6.6604476   5.83803701  1.83654857  9.25201702
   0.5908888   6.3601737   5.97806025]
 [ 2.50308657  2.73537254  7.75233459 10.91046619  8.02453804 -2.01198173
   1.12858331  2.27532029  5.94676876  5.69339657  2.36897612  9.51562595
  -0.20946389  6.78722334  5.83789873]
 [ 0.66360563 -0.07857031  5.66631746 11.29061985  4.54676151  1.32678747
   0.05444855  6.98295021  8.58084011  6.06848574  0.7207346   9.25003147
   4.34698868  7.33756542  8.16890621]
 [ 0.72273016 -0.79444987  5.70452118 11.32446957  4.15134716  1.98510754
   0.05090989  7.62190199  9.18118858  6.13288403  0.8108086   9.28542709
   5.01382971  7.22703266  8.46621895]]
[[0.         0.00295257 0.07035912 0.09047316]
 [0.00295257 0.         0.09028325 0.11269992]
 [0.07035912 0.09028325 0.         0.00171737]
 [0.09047316 0.11269992 0.00171737 0.        ]]
Skipped small class io.aiven

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[10 10 10 10  2  2  2  2  2 13 23 23 23  0 34 24 24 18 24  6  1 21 21 16
 -1  3  3  3  3 17 17 16 17 20 20 20 18 34 32 18 18 18 18 18 18 25  9  9
  9 23 26 19  9  4 29 17 26 14 14 35 35 35 32 33 32 32 33 14 17 17 16 14
 26 25 24 32 26 25 15 15 26 26 26 15  8 24 24 24 24  4 29 -1 32 26 26 26
 27 28 28 28 24  6 27 22 22 22 24 23  0  0  5 25 22 22 12 12 21 26  1  6
  6  6  6  6  6  6  6 12  1  1 11 11 35 11 11  0  0  0 19 11 11 13  4 33
 13 23 13 13 13 13 30 29 29 13 30 31 30 24 32 31 -1 29 29  4 31 29 29 31
 31 14 13 14 12 12 27 12 27 27 12 21 29 -1 18 19 19 19 19 15 19  1  1  7
  8  8  7  7  8  8  8  5  5  7  5 34 18]
[[ 3.45454764  4.60297012  5.94504738  6.90563488  7.97012949  7.38699341
   1.30478454  4.21073532  7.24094439 -0.99019378  3.88367629 -2.77729583
  -0.1111766   5.22243118  7.93957806]
 [11.55466557  1.77269936  3.09606099 12.41927624  4.8333478   1.71579278
   5.53294468  4.1751318   9.50000095  4.70260382  3.7654438   5.99496508
   0.1746695   7.61550426  4.44544411]
 

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 2 2 2 2 3 2 2 3 3 3 1 1 1 1 1 1 1 0]
[[ 0.48666865  9.86628914  9.60192776  4.02210999  2.18164945  2.86430264
   0.62448239 10.58190441  5.26127005  6.76482534  6.69706488  1.45987773
   3.81260777  1.7207011   8.62092495]
 [ 0.09354406 10.16262627  9.66347313  3.67688727  1.96835709  2.30581236
   0.05918726 11.33863926  5.56366825  7.43630981  6.58383751  1.16918886
   4.35559368  1.41000199  9.62115383]
 [ 1.63648295  7.85615063  6.64772701  4.11080074  2.55867791  5.3686409
   4.54533148  4.74209785  6.68637753  3.39312339  6.42189455  6.28641748
   4.01329803  3.20011973  4.45389509]
 [ 2.40282679  8.45636749  6.66558504  4.83107281  2.23516726  5.80605364
   4.213943    4.15483475  6.1262517   3.83016706  6.59666252  6.38426161
   4.40790367  3.62259078  4.21036053]]
[[0.         0.0022846  0.12329947 0.13285014]
 [0.0022846  0.         0.14871299 0.15993739]
 [0.12329947 0.14871299 0.         0.00371353]
 [0.13285014 0.15993739 0.00371353 0.        ]]
Skipped small class i

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[1 1 2 2 3 2 3 0 0 0 3 1]
[[ 6.2793088  10.15185165  5.91741276  1.10783017  4.02462387  5.27819538
   2.91091347  2.75811577  4.01905155  5.25614023  4.70906353  1.44544232
   4.11170578  7.68894291 10.69307041]
 [ 6.17678499 10.21238708  5.09304857  1.70025671  3.85423613  5.96325207
   3.22394872  3.05367994  4.1438899   4.60471344  3.96342063  1.71394622
   4.15813208  7.41339016 10.60580444]
 [ 6.35561466 10.59329224  5.40174818  1.73716259  3.1638093   5.80411768
   2.86952329  2.37073112  3.85542846  4.99924088  3.98641872  1.39246047
   4.68281555  6.59884596 10.75133324]
 [ 6.41437435 10.53252411  5.95900869  1.30912578  3.33008265  5.39332628
   2.66569757  2.16049671  3.76281691  5.43179989  4.45102358  1.22382724
   4.66933775  6.75098753 10.84238625]]
[[0.         0.0028362  0.00412371 0.00255416]
 [0.0028362  0.         0.00276716 0.00492798]
 [0.00412371 0.00276716 0.         0.00123707]
 [0.00255416 0.00492798 0.00123707 0.        ]]
Skipped small class io.aiven.klaw.re

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
[[11.8233223   8.28209591  2.60767365  3.78538609  2.4870038   6.27260256
   0.84166658  9.75563526  4.56214619  4.16256142 10.27509117  2.37154222
   4.11096334  6.16070032  5.09413338]]


/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 1 1 2 1 1 1 1 0 2 2 2 0]
[[ 6.70774412  3.9066999   2.22347331  3.72727442  7.24954987  3.81696868
   3.83018184  4.48724508  4.16285515  1.85138071  6.80563259  9.38348293
   3.24644375 13.25197601  4.76292515]
 [ 6.24620438  4.1903162   3.17723107  3.89188814  7.39435101  3.88493228
   4.19695711  4.88833475  4.54211187  1.12121487  6.26267958  8.92531395
   3.73393774 13.5584774   5.2188549 ]
 [ 6.88211918  4.54473972  2.14063263  3.64753437  7.43281794  4.38744593
   3.98732901  5.48868561  4.84094143  1.57393575  7.22543907  8.80948544
   3.42561841 13.52507591  6.03586674]]
[[0.         0.00288552 0.003151  ]
 [0.00288552 0.         0.00330676]
 [0.003151   0.00330676 0.        ]]
Skipped small class io.aiven.klaw.service.SchemaRegistrySyncControllerService, number of methods 5


/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 0 0 1 1 1 1 1 0]
[[ 8.46219826  9.48659897  3.97023368  7.64636278  1.50831413  6.90060043
   3.19179821  5.82585192 13.56089115  3.14361    -0.96464992  6.22756815
   7.58335972  3.87234974  5.50089359]
 [14.02167702 11.46538639  0.8744334   6.02179956  9.82245445  6.5810132
   8.79274082  8.35451031  9.60240173  7.47676325 -2.031919    8.00884342
   2.34955597  7.90621853  8.99185371]]
[[0.         0.12853897]
 [0.12853897 0.        ]]


/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 2 1 1 1 0 2 2 1 0 0 1 0]
[[1.35534632 2.44422436 8.46578789 4.77672195 1.73153555 8.3202343
  6.33811474 2.94988132 3.41877747 4.04291248 7.43880749 7.49708939
  6.04884386 6.97009087 6.02607536]
 [1.5780468  2.43286324 8.2872839  4.43434525 1.92986166 7.95687914
  6.54881907 3.66264391 3.01923561 3.65256333 6.91518736 7.15645933
  5.83057404 6.77818537 6.47191048]
 [1.53688002 2.21824217 8.56647301 4.57840967 1.69076967 8.10357094
  6.78044701 4.04621935 2.98532343 3.64760923 7.39198828 6.93012381
  6.08417749 6.37731647 6.47730875]]
[[0.         0.00176684 0.00288631]
 [0.00176684 0.         0.00088481]
 [0.00288631 0.00088481 0.        ]]


/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[1 1 2 2 2 2 2 2 0 0 1 0 0 2 2 0 0 0 0 0 2 2 2 1]
[[ 3.78179765  6.5608449  -2.63762283  1.10260034  5.99199867  4.31839371
   6.49784899  1.78436732  7.4375267   7.79067087  6.0187149   3.06569314
   1.9760648   8.21884251  1.95796764]
 [ 3.73540497  3.72874498  2.74946022  7.07489777  3.28250074  9.62861443
   2.15362382  4.19158888  1.19954705  6.68488836  2.92901683 -1.412521
   8.01504135  6.01707506  3.76027846]
 [ 3.50026202  3.8704977   2.67397952  7.91008806  3.80624032  9.56592369
   3.23746777  2.43785334  1.39376736  6.67162323  3.07110143 -1.35844958
   7.26832342  6.14014673  3.20682693]]
[[0.         0.31476839 0.28821594]
 [0.31476839 0.         0.00819844]
 [0.28821594 0.00819844 0.        ]]
Skipped small class io.aiven.klaw.repository.EnvRepo, number of methods 7
Skipped small class io.aiven.klaw.repository.UserInfoRepo, number of methods 9
Skipped small class io.aiven.klaw.repository.KwRolesPermsRepo, number of methods 7


/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[ 0  0  4  0  2  0  6 -1 11  0  0  3  5  7  5  9 10  8  7  7  8  8  1  1
  1  1  1 10 12 12 12 11 11 11 12  7  1  6  2  9  9  9  9 10  7  3  3  9
  2  2  2  2  3  3  6  9  9  9  9 10 10  6  6  4  4  5  5  3  5  3  0  7]
[[ 6.39891386  2.22030878  3.96437454  6.89969015  3.7754004   4.52696371
   2.75191474  9.27766132  1.96571827  4.57018137 -0.02942746  9.72526264
   9.46167755  5.41557312  0.12329319]
 [ 3.00840211  6.00359821  5.8127408   2.2596519   1.12010431  7.20548582
   3.957546    4.0068717   8.29391384  5.41653681  5.96351624  3.40633082
   3.39987254  7.10024977  4.5321846 ]
 [ 5.14604759  6.85795069  4.59010935  4.41266727  2.72658062  2.28003788
   5.406703    5.50743437  4.25361252  2.87894654  3.62898159  5.03041506
   2.48364568  4.84789276  7.29189587]
 [ 5.41635656  5.92941427  5.84978724  3.41834331  5.01031017  5.82625151
   7.06590796  5.80675745  4.67279196  5.71964216  5.07746601  5.1013217
   1.73249865  6.30761671  5.24687767]
 [ 4.95838451  5.95165539  4.2960

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 1 1 1 1 5 3 0 2 2 4 4 5 1 3 5 3 0 3 3 2 3 4]
[[ 6.82471609  5.95360565  5.20333338  5.24170065  5.59361267  7.1585145
   6.02777338  5.61904335  7.00391722  4.27259159  5.63700676  2.09047842
   6.52854013  7.43706274  4.02425861]
 [ 3.42956424  6.59848356  1.05519462  9.77320957  4.38031578  0.81411469
   7.53742456  5.47366047  7.99424076 10.95160007  6.71471167 -0.02706247
   6.10137796  5.02888012  4.59113312]
 [ 3.36486578  5.43215752  1.62837613  6.4807086   4.31827402  1.99464595
   7.56349468  5.97459984  6.04397726  6.67686987  7.57468939  1.29272151
   7.53968191  5.75417757  4.56615591]
 [ 3.0561769   5.42168713  1.32319522  7.07251358  4.40355968  2.04197598
   7.5823369   6.28848028  6.41099834  7.3969593   7.39616823  1.14335346
   7.27992868  5.96872759  4.24006081]
 [ 3.86693048  5.06450605  2.22050738  6.14021826  4.80517197  3.26369286
   7.36908817  6.64758873  6.51476192  6.01635695  7.06624651  1.85741448
   7.84704208  6.43508148  4.37860918]
 [ 3.68718243 

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[2 0 2 7 7 2 4 6 6 0 3 6 3 0 1 7 0 1 7 0 1 4 3 6 4 4 4 5 5 3 5 2 7]
[[ 6.48425484  5.9437027   6.62288761 -1.41878331  0.73615927  6.81867266
   0.49685246  6.452806    8.302248    5.91249323  0.89044756  5.54410505
   5.7255106   2.30808496  3.02721262]
 [ 6.44821167  6.03838778  6.96493149 -1.34708428  0.4511604   6.5502553
   0.57216704  6.37957239  8.1398983   6.3992486   0.93832707  5.63589668
   6.1424551   2.70718908  2.9840107 ]
 [ 4.90365934  6.29041719  9.21893501  4.47749805  3.61110163  7.31788874
   3.11005306  4.02551508  4.93417835  5.44717979  2.84019899  6.63562155
   4.86520863  2.37325048  5.38726807]
 [ 4.2430315   5.66677046  9.82223701  5.08257771  3.81596065  6.51788378
   2.79649925  4.11780882  4.33884764  6.25266027  2.8237071   6.72553301
   5.24040174  1.9882077   5.96925449]
 [ 5.55050611  6.9779954   8.48652649  3.70922613  3.47626734  8.15795517
   3.52248597  4.12810469  5.75704813  4.487607    3.16182351  6.67201853
   4.04191685  3.34786725  4.74396706

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[ 5  7  6  1  1  1  1  4 21 21 21 18  2  7  7  1  1 14  4  4  4 17 14 15
 22 22 14 14 18 -1 15 -1 15 21 17 18  7 21 21 18 18  7  7 14  8  9  8  8
  8 17 20 15 15  2  2  2  2  2  2  2  2  2  9 14 11 11 12 17 14 14 -1 -1
 16 22  5 22 16 16 18 18 17 18 21  9  9 21 19 19 19 20 20  8  0  0  0 10
  0 10 14 14 10 10 12 -1 12  9 12 12  9 12 13 13 13 13 13 11 11 11  8  3
  3  5  5  6  6  6 19 19  6  3  4  1]
[[ 5.05790043  2.61972904 -1.86391938  3.21524     1.24973083  6.51348782
  -1.57235622  2.62171984  8.9735117   9.9003706   8.2264843   7.39288902
   3.8508141   3.66581631  8.42873383]
 [ 3.720644    6.18698502  3.35216475  9.07546139  3.37686443  9.21731377
  -0.7456401   8.45665741  3.874331    3.41926169  5.36719847  9.0933485
   3.90835667 11.27236271  5.27360153]
 [ 5.37633371  4.90642977  4.27646971  5.6182065   4.330791    5.02964783
   3.97942281  6.17565298  4.25041389  3.7408483   5.50192547  4.64607525
   3.767205    6.87689066  4.59993792]
 [ 8.16483402  3.3314805   2.32374692

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 1 1 1 1 1 1 1 0 0 0 0]
[[ 5.36215115  8.08856773  4.33832169  4.16987038  0.88911837  4.99496698
   5.686759    8.22287369  0.0339754   0.55891687  4.65855408  5.75286007
   4.44997883  3.08154058  5.3524003 ]
 [ 5.25546074  8.63023853  4.21919823  3.78321218  1.84457242  6.07661724
   5.91770649  8.61038113 -0.09045331  0.69309503  5.28005934  5.00722265
   3.94197702  2.94179678  4.14768028]]
[[0.         0.00722355]
 [0.00722355 0.        ]]
Skipped small class io.aiven.klaw.service.ValidateCaptchaService, number of methods 1


/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[-1 -1  0  0  0  0  0  0  0  0  0  0]
[[ 5.77886391  5.18610573  6.25833559 -1.51266193  2.66629887  3.61936164
   1.64602864  5.72337294  1.62600863  2.60421848  3.52156281  3.89653492
   5.35178185  5.56787109  5.0161953 ]]
Skipped small class io.aiven.klaw.service.MailUtils.MailInfo, number of methods 6
Skipped small class io.aiven.klaw.service.KafkaConnectSyncControllerService, number of methods 6
Skipped small class io.aiven.klaw.service.interfaces.HAMessagingServiceI, number of methods 2
Skipped small class io.aiven.klaw.service.RequestStatisticsService, number of methods 1
Skipped small class io.aiven.klaw.service.EnvControllerService, number of methods 2


/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[10 10 10 10  6  6  6  6  6 13 28 28 28  2 17 31 31 23 31 14  0 22 22  1
  1  1  1 15 15 15 22 15 20 20 20 23 17 18 23 23 23 23 -1 23 19 26 26 26
 29 34 14 26  4 36 21 34 13 13 17 17 17 18 17 18 18 17 13 21 21 22 13 34
 19 -1 -1 33 19 24 24 34 34 34 24  8 30 31 30 30  4 36 -1 -1 33 34 34 37
 37 37 37 31 14 -1 27 27 27 31 29  2  2  5 19 27 27 16 16 11 11 11 33  0
 14 14 14 14 14 14 14 14 16  0  0  9  9 17  9  9  9  9 12 12 12 12 32  4
 17 13 12 36 36 12 32 35 32 31 18 35 -1 36 36  4 35 36 36 35 35 13 -1 13
 16 16 38 16 38 38 16 11 36 -1 23 25 25 25 25 25  0  0 24  8  8  3  3  3
 25  7  7  7  8  8  8  5  5  7  5 15 23 29]
[[ 6.96537018  6.46882963  5.58154392  4.72941017  1.94326556  7.5119133
   9.66941929  8.75453281 12.14411545  6.54756403 -0.74884838  1.98461843
   2.03074384 10.30067062  6.22460985]
 [ 2.97438288  7.4399066   3.54878354  5.79844141  6.63241482  4.03872967
   6.59861374  7.37341785  6.61245203  5.29786205  3.07968712  2.59734344
   3.83073592  9.12397671  3.08936501]

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[3 3 0 0 3 3 0 5 0 4 4 1 6 1 4 2 2 4 1 2 1 2 2 2 2 6 6 5 0 0 1 5 1]
[[3.86911416 2.34098458 7.77746439 3.28281021 5.63648033 7.5877676
  5.08256292 0.19646372 8.20419025 7.93464279 3.20802236 5.34823322
  8.73320484 5.35257292 3.41762161]
 [5.31451893 9.15465355 2.79711866 5.17443752 4.32009029 4.3945508
  6.3920722  4.65970802 7.01628637 7.26782656 6.99559593 1.88389826
  0.985906   3.74127412 5.17464638]
 [5.01339149 9.06581211 2.77768636 5.27957153 4.32504845 4.98919916
  6.68230104 4.76202106 6.7918272  7.18005085 7.41941023 1.64449227
  0.98862851 3.94175339 4.54770565]
 [1.33434808 3.0663476  6.77521992 2.04201603 4.17877531 8.00907135
  7.01965141 1.00305367 6.0736165  8.76089573 6.89233828 3.98504329
  7.33489513 5.0958662  1.32619894]
 [1.56833875 3.00951934 8.29024696 2.89766121 3.80574393 7.02682066
  7.28450918 1.53054667 6.27661705 7.7700386  8.08061218 2.79325271
  6.36414385 4.07688999 1.39186573]
 [1.32272828 3.010185   7.94383526 2.63604736 3.89562392 7.32587719
  7.15

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[1 1 1 0 0 1 1 0 0 1 1 2 1 3 3 0 3 3 2 3 2 2 3 3 2 2 0]
[[ 3.02881479 11.09700203  6.24696112  6.33294439  6.74410057  5.56821394
   1.55123782  6.85278034  6.3208065   2.28344035  2.13372707  2.3720839
   3.64132595  8.02781582  7.75118351]
 [ 4.59279537 10.4809761   4.57204008  5.86616707  5.59222651  6.7073822
   0.37267631  5.67723465  7.43568277  2.07442236  3.5741291   1.21475446
   4.31857157  8.27935219  7.96349716]
 [ 3.42452693  9.54158783  4.26921558  6.25769758  4.27976847  6.71418047
   1.29013026  4.32279968  7.63449097  1.62242508  4.64893055  1.99960291
   2.99063182  8.19195271  6.71907139]
 [ 2.78733778  9.26200008  4.67873049  6.55031395  3.91098952  6.95781565
   1.79621398  4.00317383  7.24702215  1.86476898  4.92618084  2.52602053
   2.45093536  7.88031101  6.74756813]]
[[0.         0.01567408 0.02791352 0.03233807]
 [0.01567408 0.         0.01018083 0.01936204]
 [0.02791352 0.01018083 0.         0.00244837]
 [0.03233807 0.01936204 0.00244837 0.        ]]
Skipped 

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 0 1 1 1 1 1 1 1 0 1]
[[ 3.17421794  6.11685085  3.94357395  4.15917349  1.90872252  1.71683943
   1.57043517  8.87600708  2.90175462 12.07382393  3.85241914  9.46307468
   3.83505416  6.66524267  8.27060795]
 [ 2.05888987  6.9467535   4.36547184  3.03988886  2.52354717  1.85784316
   3.32554221  8.91005707  1.28482163 11.15004063  3.73622942 10.55649853
   3.25428581  7.02361965  8.95743942]]
[[0.         0.01092428]
 [0.01092428 0.        ]]
Skipped small class io.aiven.klaw.service.TopicOverviewService, number of methods 1
Skipped small class io.aiven.klaw.service.PasswordService, number of methods 1
Skipped small class io.aiven.klaw.service.utils.CacheService, number of methods 9


/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[2 0 0 4 3 4 1 1 1 0 0 4 4 0 4 4 1 4 4 1 3 3 2 2 2 4 3]
[[ 4.54665947  8.66971302  7.62280226  4.93032598  0.91945106  6.49742365
   7.20365286  8.15595055  4.39495325  5.8041029   5.60829926  3.83722091
  -1.4532429   7.5949297   2.76579046]
 [ 7.63732386  0.304234    6.81150627  5.58620882  3.01805353  5.71977568
   7.17530012  6.49628973  7.47657347  8.59115505  2.33763623  8.56563473
   3.46302867  3.98564696  5.48809767]
 [ 5.86437368  2.70523834  6.09107065  5.53691006  2.20642757  6.20384407
   6.48758602  6.66128302  5.14275122  6.78819466  5.04407454  6.54684353
   0.50612384  6.05280828  4.46846199]
 [ 6.15144157  2.44085503  6.27061605  5.29343462  1.8785634   6.57508039
   6.19620085  6.83609629  5.16934443  6.68096304  4.98150015  6.5494194
   0.62214822  6.00070143  4.64225721]
 [ 5.47308302  3.5332911   6.45380545  5.16374159  2.14680195  6.19406271
   5.92977095  6.82807493  4.66071129  6.27075195  5.67838144  5.75020218
  -0.1837948   6.39112997  3.97699928]]
[[0.     

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 2 2 2 2 1 1 1 1 1 3 3 0 3 2 0 2 0 1 2 1]
[[2.22201371 5.86387682 2.17156601 6.59754324 2.30637288 6.580266
  6.80147314 6.81690836 3.76300597 6.55213165 5.1229744  4.04812193
  3.85920906 5.52163363 3.84154391]
 [3.47537589 5.879426   2.61130691 5.74028063 2.42029858 6.74161959
  6.11142111 6.50956106 3.91172743 6.76852274 4.99054337 3.29022861
  3.69345927 5.12688732 4.01368999]
 [3.42448235 6.25386333 2.87659335 5.78540039 3.12847471 5.89528799
  6.14325953 6.19224977 3.29475927 6.0675025  4.17979908 3.49082351
  4.75459719 4.77986288 3.2294476 ]
 [2.85176301 6.34125757 3.06353974 6.30485773 3.37403035 5.59015322
  6.12452602 6.38099575 3.21196342 5.77825165 3.97705555 3.90462804
  5.08851576 4.96639538 2.98365688]]
[[0.         0.00495292 0.00937734 0.01104882]
 [0.00495292 0.         0.00637519 0.01254678]
 [0.00937734 0.00637519 0.         0.00187752]
 [0.01104882 0.01254678 0.00187752 0.        ]]
Skipped small class io.aiven.klaw.service.AclSyncControllerService, number of 

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[0 0 1 1 0 1 0 0 1 0 0]
[[ 2.75592756  2.71043801  3.9216609   2.52366447  4.30775595  5.39291191
   3.65885234 10.0711937   1.73156798  4.8011713   0.26304039  4.56554222
   8.64547348  4.66996098  6.92885876]
 [ 3.12334442  3.17838359  3.72657609  2.54554129  3.73415852  4.49969006
   4.5996232   8.99904251  1.51468205  6.2363534   0.43065357  5.05084419
   8.34596157  5.40618849  6.96269989]]
[[0.         0.00839085]
 [0.00839085 0.        ]]
Skipped small class io.aiven.klaw.service.MetricsControllerService, number of methods 1
Skipped small class io.aiven.klaw.service.UiControllerLoginService, number of methods 4


/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


[ 0  1  1  0 -1 -1  1  4  2  2  2  0  0  2  3  1  1  0  3  2  3  3  3  4
  4 -1  2]
[[4.1809001  6.31211996 5.78824329 5.67609978 3.26116037 6.27664757
  4.94508982 2.76319909 2.96458054 1.4923712  6.16565943 4.8074007
  0.751266   5.46352386 6.0700469 ]
 [4.11912823 7.27353477 4.83613443 5.25317764 3.47559476 5.87574005
  5.19786215 3.66181016 3.79764318 1.50670493 6.7842617  4.16454554
  0.60573167 5.97381067 5.46079969]
 [3.98289251 8.27232838 3.83795333 5.2815814  3.32431483 5.81576061
  5.26329422 4.22853756 4.7400713  1.87605214 7.14150476 3.71958423
  0.78941518 6.34728146 4.90503025]
 [4.49100733 9.00210476 3.22273397 6.79120302 3.07565594 5.88657379
  5.57196808 4.77254009 5.5436511  2.74363875 6.45845652 3.6676054
  1.35797083 6.68660975 3.84472895]
 [4.27998924 8.79960728 3.20490026 6.57909155 3.03245068 6.19913626
  5.56889105 4.90244436 5.70437193 2.52930093 6.405159   3.41050649
  1.03661525 6.34488297 4.48252249]]
[[0.         0.00740795 0.0241373  0.04429768 0.04047387]

/home/martin/Projects/Master's Thesis/JENA/JENA.V2/.venv/lib/python3.13/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


In [94]:
# input_file = "/media/martin/BigData/datasets/github-projects/klaw/multi-service-final-multiply_signature-remove-words_yes.jsonl"
input_file = (
    "/media/martin/BigData/datasets/github-projects/klaw/final-with-centroids.jsonl"
)
# input_file = "/media/martin/BigData/datasets/pa165-git/multi-service-with-distances.jsonl"
# input_file = "/media/martin/BigData/datasets/github-projects/cas/multi-service-result-codebert-combined.jsonl"

df = pd.read_json(input_file, lines=True)
df = df.query("optimal_k.notnull()")
more_cluster_lines = df.query("optimal_k > 2")
one_cluster_lines = df.drop(more_cluster_lines.index)


def count_rows_and_print(df: pd.DataFrame, query: str, stat_type: str):
    df_part = df.query(query)
    count = len(df_part)
    print(f"{stat_type}: {count}")
    for _, record in df_part.iterrows():
        print(record["fullyQualifiedName"])
        # print(json.dumps(record["methods"], indent=2))
    print("---")
    return count


false_negatives = count_rows_and_print(
    one_cluster_lines, "label == 1", "False negatives"
)
true_negatives = count_rows_and_print(one_cluster_lines, "label == 0", "True negatives")
false_positives = count_rows_and_print(
    more_cluster_lines, "label == 0", "False positives"
)
true_positives = count_rows_and_print(
    more_cluster_lines, "label == 1", "True positives"
)

precision = true_positives / (true_positives + false_positives)
recall = true_positives / (true_positives + false_negatives)
accuracy = (true_positives + true_negatives) / (
    true_positives + false_positives + false_negatives + true_negatives
)
f1_score = 2 * ((precision * recall) / (precision + recall))

print(f"""
Stats
Precision: {precision}
Recall: {recall}
Accuracy: {accuracy}
F1 score: {f1_score}
-------------
""")

# eval_df = df[df["label"].notna()]
# y_true = eval_df["label"].astype(int)
# y_pred = pd.Series(0, index=eval_df.index)
# y_pred[eval_df.index.isin(more_cluster_lines.index)] = 1

# print("Stats")
# print(classification_report(y_true, y_pred, target_names=["Cohesive", "Multiservice"]))

# class_clusters = defaultdict(dict)
#
# for _, record in more_cluster_lines.iterrows():
#     k = record["optimal_k"]
#     clusters = record["cluster"]
#     methods = record["methods"]
#     class_name = record["fullyQualifiedName"]
#     cohesion = record["lcom4"]
#     noise_ratio = record["noise_ratio"]
#
#     method_clusters = {}
#     for index, cluster in enumerate(clusters):
#         cluster_number_str = str(cluster)
#         method = methods[index]
#         if cluster_number_str in method_clusters:
#             method_clusters[cluster_number_str].append(method)
#         else:
#             method_clusters[cluster_number_str] = [method]
#
#     class_clusters[class_name]["methods"] = method_clusters
#     class_clusters[class_name]["lcom4"] = cohesion
#     class_clusters[class_name]["noise_ratio"] = noise_ratio
#
# for className, metadata in class_clusters.items():
#     print(f"--- Class {className}, LCOM4: {metadata['lcom4']} ---")
#     for cluster, methods in metadata["methods"].items():
#         print(f"Cluster {cluster}: {json.dumps(methods, indent=2)}\n")
#     print()

more_cluster_lines.query("label == 0")

False negatives: 4
io.aiven.klaw.clusterapi.utils.ClusterApiUtils
io.aiven.klaw.clusterapi.controller.ClusterApiController
io.aiven.klaw.service.ServerConfigService
io.aiven.klaw.service.UtilControllerService
---
True negatives: 9
io.aiven.klaw.clusterapi.services.ConfluentCloudApiService
io.aiven.klaw.MockMethods
io.aiven.klaw.controller.SchemaRegistryController
io.aiven.klaw.repository.MessageSchemaRepo
io.aiven.klaw.controller.AclController
io.aiven.klaw.controller.TopicController
io.aiven.klaw.repository.RegisterInfoRepo
io.aiven.klaw.service.AnalyticsControllerService
io.aiven.klaw.service.MailUtils
---
False positives: 10
io.aiven.klaw.repository.AclRequestsRepo
io.aiven.klaw.repository.AclRepo
io.aiven.klaw.repository.ActivityLogRepo
io.aiven.klaw.repository.KwKafkaConnectorRequestsRepo
io.aiven.klaw.repository.TopicRequestsRepo
io.aiven.klaw.repository.KwKafkaConnectorRepo
io.aiven.klaw.repository.SchemaRequestRepo
io.aiven.klaw.controller.KafkaConnectController
io.aiven.klaw.r

,methods,lcom4,fullyQualifiedName,startLine,label,optimal_k,noise_ratio,cluster,avg_dist
223,"[{'name': 'findById', 'signature': 'Optional<A...",19,io.aiven.klaw.repository.AclRequestsRepo,12,0,3.0,0.0,"[1, 1, 2, 2, 2, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, ...",3.557154
335,"[{'name': 'findById', 'signature': 'Optional<A...",24,io.aiven.klaw.repository.AclRepo,11,0,5.0,0.0,"[1, 3, 4, 2, 2, 3, 4, 2, 4, 1, 2, 2, 3, 2, 0, ...",7.585466
355,"[{'name': 'findById', 'signature': 'Optional<A...",11,io.aiven.klaw.repository.ActivityLogRepo,12,0,3.0,0.0,"[1, 1, 2, 2, 2, 2, 0, 0, 1, 0, 2]",0.483613
386,"[{'name': 'findById', 'signature': 'Optional<K...",15,io.aiven.klaw.repository.KwKafkaConnectorReque...,13,0,4.0,0.0,"[0, 3, 3, 3, 3, 3, 0, 1, 2, 1, 2, 2, 1, 1, 1, 0]",6.141423
395,"[{'name': 'findById', 'signature': 'Optional<T...",17,io.aiven.klaw.repository.TopicRequestsRepo,14,0,4.0,0.0,"[0, 0, 2, 2, 2, 2, 3, 2, 2, 3, 3, 3, 1, 1, 1, ...",9.513302
402,"[{'name': 'findById', 'signature': 'Optional<K...",12,io.aiven.klaw.repository.KwKafkaConnectorRepo,11,0,4.0,0.0,"[1, 1, 2, 2, 3, 2, 3, 0, 0, 0, 3, 1]",0.307438
409,"[{'name': 'findById', 'signature': 'Optional<S...",13,io.aiven.klaw.repository.SchemaRequestRepo,12,0,3.0,0.0,"[0, 1, 1, 2, 1, 1, 1, 1, 0, 2, 2, 2, 0]",0.311443
412,"[{'name': 'createConnectorRequest', 'signature...",1,io.aiven.klaw.controller.KafkaConnectController,32,0,3.0,0.0,"[0, 2, 1, 1, 1, 0, 2, 2, 1, 0, 0, 1, 0]",0.184599
413,"[{'name': 'findById', 'signature': 'Optional<T...",24,io.aiven.klaw.repository.TopicRepo,11,0,3.0,0.0,"[1, 1, 2, 2, 2, 2, 2, 2, 0, 0, 1, 0, 0, 2, 2, ...",20.372759
445,"[{'name': 'getClusterApiStatus', 'signature': ...",1,io.aiven.klaw.service.ClusterApiService,99,0,4.0,0.0,"[0, 0, 2, 2, 2, 2, 1, 1, 1, 1, 1, 3, 3, 0, 3, ...",0.769643


In [ ]:
from pipelines.shared import input_until_integer

# Labeling script

input_file = "/media/martin/BigData/datasets/github-projects/klaw/multi-service.jsonl"
output_file = (
    "/media/martin/BigData/datasets/github-projects/klaw/multi-service-labeled.jsonl"
)

df = read_json(input_file, lines=True)

df["label"] = pd.Series()

for index, row in df.iterrows():
    name = row["fullyQualifiedName"]
    methods = row["methods"]
    assigned_label = row["label"]
    lcom4 = row["lcom4"]

    label = 0
    if not should_be_skipped(name) and len(methods) > 10:
        print("-" * 10)
        print(f"Class {name} (LCOM4 {lcom4})")
        print("Methods: ")
        for method in methods:
            print(method)

        label = input_until_integer("Enter label: ")

    df.loc[index, "label"] = label
    df.to_json(output_file, lines=True, orient="records")

# Results

## Remove words: yes, names multiply: 1.3

```
False negatives: 1
io.aiven.klaw.service.UtilControllerService
---
True negatives: 10
io.aiven.klaw.MockMethods
io.aiven.klaw.repository.MessageSchemaRepo
io.aiven.klaw.repository.ActivityLogRepo
io.aiven.klaw.controller.AclController
io.aiven.klaw.repository.KwKafkaConnectorRepo
io.aiven.klaw.controller.TopicController
io.aiven.klaw.repository.SchemaRequestRepo
io.aiven.klaw.controller.KafkaConnectController
io.aiven.klaw.service.AnalyticsControllerService
io.aiven.klaw.service.MailUtils
---
False positives: 9
io.aiven.klaw.clusterapi.services.ConfluentCloudApiService
io.aiven.klaw.controller.SchemaRegistryController
io.aiven.klaw.repository.AclRequestsRepo
io.aiven.klaw.repository.AclRepo
io.aiven.klaw.repository.KwKafkaConnectorRequestsRepo
io.aiven.klaw.repository.TopicRequestsRepo
io.aiven.klaw.repository.RegisterInfoRepo
io.aiven.klaw.repository.TopicRepo
io.aiven.klaw.service.ClusterApiService
---
True positives: 15
io.aiven.klaw.clusterapi.utils.ClusterApiUtils
io.aiven.klaw.clusterapi.controller.ClusterApiController
io.aiven.klaw.UtilMethods
io.aiven.klaw.helpers.HandleDbRequests
io.aiven.klaw.controller.TemplateMapController
io.aiven.klaw.controller.UsersTeamsController
io.aiven.klaw.controller.EnvsClustersTenantsController
io.aiven.klaw.helpers.db.rdbms.SelectDataJdbc
io.aiven.klaw.helpers.db.rdbms.HandleDbRequestsJdbc
io.aiven.klaw.helpers.db.rdbms.InsertDataJdbc
io.aiven.klaw.helpers.db.rdbms.DeleteDataJdbc
io.aiven.klaw.service.ServerConfigService
io.aiven.klaw.service.UsersTeamsControllerService
io.aiven.klaw.service.CommonUtilsService
io.aiven.klaw.clusterapi.UtilMethods
---

Stats
Precision: 0.625
Recall: 0.9375
Accuracy: 0.7142857142857143
F1 score: 0.75
```

## Remove words: yes, names multiply: no

```
False negatives: 2
io.aiven.klaw.clusterapi.controller.ClusterApiController
io.aiven.klaw.service.UtilControllerService
---
True negatives: 13
io.aiven.klaw.clusterapi.services.ConfluentCloudApiService
io.aiven.klaw.MockMethods
io.aiven.klaw.controller.SchemaRegistryController
io.aiven.klaw.repository.MessageSchemaRepo
io.aiven.klaw.repository.ActivityLogRepo
io.aiven.klaw.controller.AclController
io.aiven.klaw.repository.KwKafkaConnectorRepo
io.aiven.klaw.controller.TopicController
io.aiven.klaw.repository.SchemaRequestRepo
io.aiven.klaw.controller.KafkaConnectController
io.aiven.klaw.service.AnalyticsControllerService
io.aiven.klaw.service.MailUtils
io.aiven.klaw.service.ClusterApiService
---
False positives: 6
io.aiven.klaw.repository.AclRequestsRepo
io.aiven.klaw.repository.AclRepo
io.aiven.klaw.repository.KwKafkaConnectorRequestsRepo
io.aiven.klaw.repository.TopicRequestsRepo
io.aiven.klaw.repository.RegisterInfoRepo
io.aiven.klaw.repository.TopicRepo
---
True positives: 14
io.aiven.klaw.clusterapi.utils.ClusterApiUtils
io.aiven.klaw.UtilMethods
io.aiven.klaw.helpers.HandleDbRequests
io.aiven.klaw.controller.TemplateMapController
io.aiven.klaw.controller.UsersTeamsController
io.aiven.klaw.controller.EnvsClustersTenantsController
io.aiven.klaw.helpers.db.rdbms.SelectDataJdbc
io.aiven.klaw.helpers.db.rdbms.HandleDbRequestsJdbc
io.aiven.klaw.helpers.db.rdbms.InsertDataJdbc
io.aiven.klaw.helpers.db.rdbms.DeleteDataJdbc
io.aiven.klaw.service.ServerConfigService
io.aiven.klaw.service.UsersTeamsControllerService
io.aiven.klaw.service.CommonUtilsService
io.aiven.klaw.clusterapi.UtilMethods
---

Stats
Precision: 0.7
Recall: 0.875
Accuracy: 0.7714285714285715
F1 score: 0.7777777777777777
-------------
```

## Remove words: no, names multiply: no

```
False negatives: 2
io.aiven.klaw.clusterapi.controller.ClusterApiController
io.aiven.klaw.service.UtilControllerService
---
True negatives: 10
io.aiven.klaw.clusterapi.services.ConfluentCloudApiService
io.aiven.klaw.MockMethods
io.aiven.klaw.repository.MessageSchemaRepo
io.aiven.klaw.controller.AclController
io.aiven.klaw.repository.KwKafkaConnectorRepo
io.aiven.klaw.controller.TopicController
io.aiven.klaw.controller.KafkaConnectController
io.aiven.klaw.service.AnalyticsControllerService
io.aiven.klaw.service.MailUtils
io.aiven.klaw.service.ClusterApiService
---
False positives: 9
io.aiven.klaw.controller.SchemaRegistryController
io.aiven.klaw.repository.AclRequestsRepo
io.aiven.klaw.repository.AclRepo
io.aiven.klaw.repository.ActivityLogRepo
io.aiven.klaw.repository.KwKafkaConnectorRequestsRepo
io.aiven.klaw.repository.TopicRequestsRepo
io.aiven.klaw.repository.SchemaRequestRepo
io.aiven.klaw.repository.RegisterInfoRepo
io.aiven.klaw.repository.TopicRepo
---
True positives: 14
io.aiven.klaw.clusterapi.utils.ClusterApiUtils
io.aiven.klaw.UtilMethods
io.aiven.klaw.helpers.HandleDbRequests
io.aiven.klaw.controller.TemplateMapController
io.aiven.klaw.controller.UsersTeamsController
io.aiven.klaw.controller.EnvsClustersTenantsController
io.aiven.klaw.helpers.db.rdbms.SelectDataJdbc
io.aiven.klaw.helpers.db.rdbms.HandleDbRequestsJdbc
io.aiven.klaw.helpers.db.rdbms.InsertDataJdbc
io.aiven.klaw.helpers.db.rdbms.DeleteDataJdbc
io.aiven.klaw.service.ServerConfigService
io.aiven.klaw.service.UsersTeamsControllerService
io.aiven.klaw.service.CommonUtilsService
io.aiven.klaw.clusterapi.UtilMethods
---

Stats
Precision: 0.6086956521739131
Recall: 0.875
Accuracy: 0.6857142857142857
F1 score: 0.717948717948718
-------------
```

## Remove words: no, names multiply: 1.3
```
False negatives: 2
io.aiven.klaw.clusterapi.controller.ClusterApiController
io.aiven.klaw.service.UtilControllerService
---
True negatives: 9
io.aiven.klaw.clusterapi.services.ConfluentCloudApiService
io.aiven.klaw.MockMethods
io.aiven.klaw.repository.ActivityLogRepo
io.aiven.klaw.controller.AclController
io.aiven.klaw.repository.KwKafkaConnectorRepo
io.aiven.klaw.controller.TopicController
io.aiven.klaw.controller.KafkaConnectController
io.aiven.klaw.service.AnalyticsControllerService
io.aiven.klaw.service.MailUtils
---
False positives: 10
io.aiven.klaw.controller.SchemaRegistryController
io.aiven.klaw.repository.AclRequestsRepo
io.aiven.klaw.repository.MessageSchemaRepo
io.aiven.klaw.repository.AclRepo
io.aiven.klaw.repository.KwKafkaConnectorRequestsRepo
io.aiven.klaw.repository.TopicRequestsRepo
io.aiven.klaw.repository.SchemaRequestRepo
io.aiven.klaw.repository.RegisterInfoRepo
io.aiven.klaw.repository.TopicRepo
io.aiven.klaw.service.ClusterApiService
---
True positives: 14
io.aiven.klaw.clusterapi.utils.ClusterApiUtils
io.aiven.klaw.UtilMethods
io.aiven.klaw.helpers.HandleDbRequests
io.aiven.klaw.controller.TemplateMapController
io.aiven.klaw.controller.UsersTeamsController
io.aiven.klaw.controller.EnvsClustersTenantsController
io.aiven.klaw.helpers.db.rdbms.SelectDataJdbc
io.aiven.klaw.helpers.db.rdbms.HandleDbRequestsJdbc
io.aiven.klaw.helpers.db.rdbms.InsertDataJdbc
io.aiven.klaw.helpers.db.rdbms.DeleteDataJdbc
io.aiven.klaw.service.ServerConfigService
io.aiven.klaw.service.UsersTeamsControllerService
io.aiven.klaw.service.CommonUtilsService
io.aiven.klaw.clusterapi.UtilMethods
---

Stats
Precision: 0.5833333333333334
Recall: 0.875
Accuracy: 0.6571428571428571
F1 score: 0.7000000000000001
```

## Remove words: yes, signature multiply: 1.3

```
False negatives: 1
io.aiven.klaw.clusterapi.controller.ClusterApiController
---
True negatives: 10
io.aiven.klaw.clusterapi.services.ConfluentCloudApiService
io.aiven.klaw.MockMethods
io.aiven.klaw.controller.SchemaRegistryController
io.aiven.klaw.repository.MessageSchemaRepo
io.aiven.klaw.repository.KwKafkaConnectorRepo
io.aiven.klaw.controller.TopicController
io.aiven.klaw.repository.SchemaRequestRepo
io.aiven.klaw.controller.KafkaConnectController
io.aiven.klaw.service.MailUtils
io.aiven.klaw.service.ClusterApiService
---
False positives: 9
io.aiven.klaw.repository.AclRequestsRepo
io.aiven.klaw.repository.AclRepo
io.aiven.klaw.repository.ActivityLogRepo
io.aiven.klaw.controller.AclController
io.aiven.klaw.repository.KwKafkaConnectorRequestsRepo
io.aiven.klaw.repository.TopicRequestsRepo
io.aiven.klaw.repository.RegisterInfoRepo
io.aiven.klaw.repository.TopicRepo
io.aiven.klaw.service.AnalyticsControllerService
---
True positives: 15
io.aiven.klaw.clusterapi.utils.ClusterApiUtils
io.aiven.klaw.UtilMethods
io.aiven.klaw.helpers.HandleDbRequests
io.aiven.klaw.controller.TemplateMapController
io.aiven.klaw.controller.UsersTeamsController
io.aiven.klaw.controller.EnvsClustersTenantsController
io.aiven.klaw.helpers.db.rdbms.SelectDataJdbc
io.aiven.klaw.helpers.db.rdbms.HandleDbRequestsJdbc
io.aiven.klaw.helpers.db.rdbms.InsertDataJdbc
io.aiven.klaw.helpers.db.rdbms.DeleteDataJdbc
io.aiven.klaw.service.ServerConfigService
io.aiven.klaw.service.UsersTeamsControllerService
io.aiven.klaw.service.UtilControllerService
io.aiven.klaw.service.CommonUtilsService
io.aiven.klaw.clusterapi.UtilMethods
---

Stats
Precision: 0.625
Recall: 0.9375
Accuracy: 0.7142857142857143
F1 score: 0.75
```
